src/main/scala/is/hail/expr/ir/ExtractAggregators.scala

```diff
@@ -70,28 +70,37 @@ object ExtractAggregators {
   }
 
   private def newAggregator(ir: ApplyAggOp): RegionValueAggregator = ir match {
-    case x@ApplyAggOp(a, constructorArgs, _, aggSig) =>
-      val fb = EmitFunctionBuilder[Region, RegionValueAggregator]
-      var codeConstructorArgs = constructorArgs.map(Emit.toCode(_, fb, 1))
-
-      aggSig match {
-        case AggSignature(Collect() | Take() | CollectAsSet(), _, _, Seq(t@(_: TBoolean | _: TInt32 | _: TInt64 | _: TFloat32 | _: TFloat64 | _: TCall))) =>
-        case AggSignature(Collect() | Take() | CollectAsSet(), _, _, Seq(t)) =>
-          codeConstructorArgs ++= FastIndexedSeq(EmitTriplet(Code._empty, const(false), fb.getType(t)))
-        case AggSignature(Counter(), _, _, Seq(t@(_: TBoolean))) =>
-        case AggSignature(Counter(), _, _, Seq(t)) =>
-          codeConstructorArgs = FastIndexedSeq(EmitTriplet(Code._empty, const(false), fb.getType(t)))
-        case AggSignature(TakeBy(), _, _, Seq(aggType, keyType)) =>
-          codeConstructorArgs ++= FastIndexedSeq(EmitTriplet(Code._empty, const(false), fb.getType(aggType)),
-            EmitTriplet(Code._empty, const(false), fb.getType(keyType)))
-        case _ =>
-      }
-
-      fb.emit(Code(
-        Code(codeConstructorArgs.map(_.setup): _*),
-        AggOp.get(aggSig)
-          .stagedNew(codeConstructorArgs.map(_.v).toArray, codeConstructorArgs.map(_.m).toArray)))
-
-      Region.scoped(fb.result()()(_))
+    case x@ApplyAggOp(a, constructorArgs, _, _) =>
+
+      def getAggregator(op: AggOp, aggSig: AggSignature): RegionValueAggregator = op match {
+          case Keyed(subop) =>
+            val newAggSig = aggSig.copy(op = subop, seqOpArgs = aggSig.seqOpArgs.drop(1))
```

Comment:
Here and in `AggOp.scala` we're manipulating these `seqOpArgs` and reconstructing `AggSignature`s, which seems correct, AFAICT. However, it seems more straightforward for a `Keyed` AggOp to contain an `AggSignature`. Then there's no need to have both an `AggOp` argument to `getAggregator` and a `AggSignature` argument because an `AggSignature` is one-to-one with an `AggOp` (as is the case in master). I think `getAggregator` becomes:

```scala
def getAggregator(sig: AggSignature): RegionValueAggregator = sig match {
  case AggSignature(Keyed(childSig), Seq(), None, Seq(keyType)) =>
    KeyedRegionValueAggregator(getAggregator(childSig), keyType)
  case _ =>
    ...
```

The recursive calls in `AggOp.scala` would seem to get more streamlined as well.

It does make the parser a little weirder. The `agg_signature` parser's `ir_agg_op` would need to be an alternation where you either get a simple aggsig or a `"Keyed" ~> agg_signature`. I'm not sure exactly what the imagined python interface is but if it's something like:
```python
mt.aggregate_rows(foo =
  hl.agg.grouped(mt.super_pop, hl.agg.grouped(mt.pop, hl.agg.mean(mt.GT.n_alt_alleles()))))
```
the naive recursive method to generate a string would lend it self well to generating something like:
```
(Keyed
  (Keyed
    (Mean () None (Int))
    ()
    None
    (String))
  ()
  None
  (String))
```

Reply:
I think I like my way better. The complexity reduction in one place is being gained in another. I prefer to keep the AggOp signatures as consistent as possible. The aggSignature is already on both the ApplyAggOp and the SeqOp, which are guaranteed to be the same. I think having it in a third place on only one AggOp is confusing. We can discuss if you feel strongly.